storage/integration_test.go

```diff
@@ -3233,7 +3215,100 @@ func TestIntegration_WriterAppend(t *testing.T) {
-				// Disabled due to b/408373388; unskip after this is resolved.
+				// Disabled due to b/395944605; unskip after this is resolved.
+				// gotBytes := h.mustRead(obj)
+				// if !bytes.Equal(gotBytes, tc.content) {
+				// 	t.Errorf("content mismatch: got %v bytes, want %v bytes", len(gotBytes), len(tc.content))
+				// }
+
+				// Check object exists and Finalized attribute set as expected.
+				attrs := h.mustObjectAttrs(obj)
+				if tc.finalize && attrs.Finalized.IsZero() {
+					t.Errorf("got unfinalized object, want finalized")
+				}
+				if !tc.finalize && !attrs.Finalized.IsZero() {
+					t.Errorf("got object finalized at %v, want unfinalized", attrs.Finalized)
+				}
+
+			})
+		}
+	})
+}
+
+// Writer test for append takeover of unfinalized object.
+func TestIntegration_WriterAppendTakeover(t *testing.T) {
+	t.Skip("b/402283880")
+	ctx := skipAllButBidi(context.Background(), "ZB test")
+	multiTransportTest(ctx, t, func(t *testing.T, ctx context.Context, _, prefix string, client *Client) {
+		h := testHelper{t}
+		bucketName := prefix + uidSpace.New()
+		bkt := client.Bucket(bucketName)
+
+		h.mustCreateZonalBucket(bkt, testutil.ProjID())
+		defer h.mustDeleteBucket(bkt)
+
+		testCases := []struct {
+			name           string
+			finalize       bool
+			content        []byte
+			chunkSize      int
+			takeoverOffset int64
+		}{
+			{
+				name:           "first message takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: MiB,
+			},
+			{
+				name:           "first chunk takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: 3 * MiB,
+			},
+			{
+				name:           "middle chunk takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: 6 * MiB,
+			},
+			{
+				name:           "final chunk takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: 8*MiB + 100,
+			},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				// Create non-finalized appendable writer and upload first part of content.
+				obj := bkt.Object(tc.name + uidSpace.New()).Retryer(WithPolicy(RetryAlways))
+				defer h.mustDeleteObject(obj)
+				w := obj.If(Conditions{DoesNotExist: true}).NewWriter(ctx)
+				w.Append = true
+				w.FinalizeOnClose = false
+				w.ChunkSize = tc.chunkSize
+
+				h.mustWrite(w, tc.content[:tc.takeoverOffset])
+
+				// Takeover and write remainder of content.
+				gen := h.mustObjectAttrs(obj).Generation
+				w2, off, err := obj.Generation(gen).NewWriterFromAppendableObject(ctx)
+				if err != nil {
+					t.Fatalf("NewWriterFromAppendableObject: %v", err)
+				}
+				if off != tc.takeoverOffset {
+					t.Errorf("takeover offset: got %v, want %v", off, tc.takeoverOffset)
+				}
+				w2.FinalizeOnClose = true
```

Comment:
Should we test setting params on the writer?

Reply:
Discussed offline, yes we should.

Reply:
Added some tests with a bit more of this.